Switch to State - Label dependency

ec opened the pull request.

Commit:
Flexible state - labels flow

* Add data json field to user model
* Add activation_requirements
* Add state_triggers
* Add session opening on signup
* Allow pending users to pass AuthZ
* Switch to state-label dependency and related
* Rename discarded to deleted, remove old code
* Remove keys limits, limit data to 1024 chars, add locked and new yml format
* Remove extended, fix specs, fix ALL state calculating, change yml format
* Open session on email confirmation
* Change data position in DB, add validate_length_of
* Improve test coverage

mod commented:
- [x] Remove discared and put locked
- [x] Label are not limited to :verified we can define scope=system
- [x] move data field and allow json and max 1MB
- [x] Publish data in entity, remove extended user profile, default api is always extended
- [x] locked example label suspicious:trade suspicious:deposit

mod commented:
```
user_state:
  requirements:
    active:
      email: 'verified'
      phone: 'verified'
  mapping:
    locked:
       suspicious: 'trade'
       suspicious: 'deposit'
```

mod commented:
```
user_state:
  active: 
    all:
      - email: 'verified'
      - phone: 'verified'
   locked:
     any:
       - suspicious: 'trade'
       - suspicious: 'deposit'
```

ec commented:
I suggest to have a module structure:
   module all
      state
      state
  module any
      state
      state

Looks more readable in code and not so bad in yml
```
user_state:
  all:
    active:
      email: 'verified'
      phone: 'verified'
    banned:
      ban: 'weneedtochoose'
    deleted:
      deleted: 'weneedtochoose'
  any:
    locked:
      trade: 'suspicious'
      deposit: 'suspicious'
```

mod commented on app/api/v2/admin/users.rb:
```diff
@@ -98,6 +98,11 @@ def search(field, value)
             error!({ errors: ['admin.user.update_himself'] }, 422) if target_user.uid == current_user.uid
 
+            if update_param_key == 'state' && update_param_value == 'banned'
+              target_user.add_private_label('ban', 'weneedtochoose')
```
I'd rather not chooose; why is this necessary ?

mod commented on config/user_state_requirements.yml:
```diff
@@ -0,0 +1,23 @@
+    # labels required for user to become locked
+    locked:
+      suspicious: 'trade'
+      suspicious: 'deposit'
```
config/barong.yml
```
activation_requirements:
      email: 'verified'
      phone: 'verified'
state_triggers:
    banned:
      - ban
      - fraud
    deleted:
      - delete
    locked:
      - suspicious
      - lock
feature_xyz:
  settings:
     sdfkljdlkfgj: 24
```

later:
barong.state_triggers.banned = [ban, fraud]
app_name . component . key = value

fraud:trade
fraud:deposit
fraud:*

mod commented on app/api/v2/resource/users.rb:
```diff
@@ -41,8 +41,8 @@ def verify_otp!
 
-          current_user.update(state: 'discarded')
-          EventAPI.notify('system.user.account.discarded', current_user.as_json_for_event_api)
+          current_user.labels.create(key: 'deleted', value: 'by_user', scope: 'private')
```
delete:by_user

mod commented on app/models/user.rb:
```diff
@@ -13,8 +13,10 @@ class User < ApplicationRecord
+  validates_length_of :data, :maximum => 1024
   validate :role_exists
   validate :referral_exists
+  validates :data, data_prefix: true
```
remove data_prefix: true

mod commented on app/models/user.rb:
```diff
@@ -88,9 +83,24 @@ def update_level
+        @resulting_state = 'active'
+      end
+    end
+
```
#FIXME BarongConfig should be a feature of Barong::App
A good name for Barong::App can be Barong::Config and we can add database support or vault based configs.

mod commented on config/initializers/barong_config.rb:
```diff
@@ -0,0 +1,20 @@
+class BarongConfig
```
FIXME Move to barong.rb later

mod reviewed (changes requested):
almost ready for merge
